Source/MQTTnet/Internal/AsyncReadWriteLock.cs

```diff
@@ -0,0 +1,214 @@
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(AsyncReadWriteLock));
+            }
+
+            lock (_syncRoot)
+            {
+                if (!_isLockedForWrite && !_writeWaiters.Any())
+                {
+                    _readLockCount++;
+                    return _readCompletedTask;
+                }
+
+                var waiter = new ReadWriteLockWaiter(cancellationToken);
+                _readWaiters.Enqueue(waiter);
+
+                return waiter.Task;
+            }
+        }
+
+        public Task<IDisposable> EnterWriteAsync(CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(AsyncReadWriteLock));
+            }
+
+            lock (_syncRoot)
+            {
+                if (!_isLockedForWrite && _readLockCount <= 0)
+                {
+                    _isLockedForWrite = true;
+                    return _writeCompletedTask;
+                }
+
+                var waiter = new ReadWriteLockWaiter(cancellationToken);
+                _writeWaiters.Enqueue(waiter);
+
+                return waiter.Task;
+            }
+        }
+
+        private void ReleaseRead()
+        {
+            lock (_syncRoot)
+            {
+                if (_isDisposed)
+                {
+                    return;
+                }
+
+                _readLockCount = Math.Max(_readLockCount - 1, 0);
+
+                ApproveNextWaiter();
+            }
+        }
+
+        private void ReleaseWrite()
+        {
+            lock (_syncRoot)
+            {
+                if (_isDisposed)
+                {
+                    return;
+                }
+
+                _isLockedForWrite = false;
+
+                ApproveNextWaiter();
+            }
+        }
+
+        private void ApproveNextWaiter()
+        {
+            while (_writeWaiters.Any())
+            {
+                var waiter = _writeWaiters.Dequeue();
+                var isApproved = waiter.Approve(_writeReleaser);
+                waiter.Dispose();
+
+                if (isApproved)
+                {
+                    _isLockedForWrite = true;
+                    return;
+                }
+            }
+            while (_readWaiters.Any())
+            {
+                var waiter = _readWaiters.Dequeue();
+                var isApproved = waiter.Approve(_readReleaser);
+                waiter.Dispose();
+
+                if (isApproved)
+                {
+                    _readLockCount++;
+                    return;
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            _isDisposed = true;
```

Comment:
Should this be volatile write / read? (Except where in a `lock`)

Reply:
I'm not sure what you mean by this. Could you clarify?

Reply:
Is there a possibility that `Dispose` is called from multiple threads? I don't know.
Is there a possibility that the value of `_isDisposed` is read from multiple threads? Yes.

At least the latter case may introduce a visibility issue on weaker memory model targets (like arm).
So I think here it should be 
```suggestion
            Volatile.Write(ref _isDisposed, true);
```
and in all reads that are not within a `lock` it should be `Volatile.Read(ref _isDisposed)` to be on the safe side.

On targets with strong memory models (like x86) these volatile read/writes evaporate and have no cost.

Reply:
Thank you for pointing that out. I've been working in .NET for a very long time, and I had no idea these volatile keywords existed. You learn something new every day. Makes sense for ARM-based systems though. Thanks again.